### PR TITLE
arch: riscv: remove unneeded context switch to gp register

### PR DESCRIPTION
**Pull request by cwshu**
RISC-V global pointer (GP) register is neither caller nor callee register and it's a constant value in the single ELF file.
Thus, we don't need to save/restore GP register at ISR enter/exit. Remove it to optimize context switch performance.

Also, this patch can simply the RISC-V GP support to Zephyr. We could only initialize gp value at program start, w/o initializing it again at each thread init because GP register isn't saved.
I saw #36180 is going to add GP support so this patch will affect it.

### CODE REVIEW

**Review by katsuster — APPROVED** [review]

**Comment by cwshu** [comment]
Add another reference:
FreeRTOS RISC-V doesn't save GP/TP register because they are constants in it. 
https://github.com/FreeRTOS/FreeRTOS-Kernel/blob/V10.4.4/portable/GCC/RISC-V/portASM.S#L380

Zephyr RISC-V will set TP register to the TLS of thread, but GP register is still a constant so it doesn't need context switch.

**Comment by cwshu** [comment]
Sorry for asking, but could anyone help review this patch? It still needs one approval review.
@cfriedt @galak

**Comment by cfriedt** [comment]
> Sorry for asking, but could anyone help review this patch? It still needs one approval review.
> @cfriedt @galak

I think I will defer to someone else who has touched this file in the past.

**Comment by cwshu** [comment]
Revert commit 7b09d031fae3969fbe8647a57190f1fa21fabd8e from #36180 because GP initialization at each thread init is NOT needed after this PR removes the context save/restore of GP register.

Depends on the discussion with @uLipe in #36180, reverting this commit still works and is tested on esp32c3 soc.

**Review by uLipe — APPROVED** [review]

**Pull request merged**
